### PR TITLE
statistics: support indexes containing virtual column for full sampling

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Problem Summary:

This PR collects the NDV/FM-Sketch/nulls for indexes containing virtual columns and prefix column.

### What is changed and how it works?

Proposal: [xxx](url) 

How it Works:

We send sub index task for each index containing virtual column to get their fm-sketches to calculate 
NDV and maintain NDV for partition table.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note

### CODE REVIEW

File: executor/analyze.go [hunk 1]
```diff
@@ -524,19 +582,35 @@ func analyzeColumnsPushdown(colExec *AnalyzeColumnsExec) []analyzeResult {
...
-	if colExec.analyzeVer == statistics.Version3 {
-		count, hists, topns, fmSketches, extStats, err := colExec.buildSamplingStats(ranges, collExtStats)
+	if colExec.StatsVersion == statistics.Version3 {
+		indexesWithVirtualCol := make([]*model.IndexInfo, 0, len(colExec.indexes))
+		indexesWithVirtualColOffsets := make([]int, 0, len(colExec.indexes))
+		for i, idx := range colExec.indexes {
+			containVirtual := false
+			for _, col := range idx.Columns {
+				if colExec.colsInfo[col.Offset].IsGenerated() && !colExec.colsInfo[col.Offset].GeneratedStored {
+					containVirtual = true
```
Comment (eurekaka) [inline executor/analyze.go:592]: add an early break?

File: executor/analyze.go [hunk 2]
```diff
@@ -731,12 +819,68 @@ func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtS
...
+		for _, sample := range rootRowCollector.Samples {
+			for i := range sample.Columns {
+				sample.Columns[i], err = tablecodec.DecodeColumnValue(sample.Columns[i].GetBytes(), &e.colsInfo[i].FieldType, sc.TimeZone)
+				if err != nil {
+					return 0, nil, nil, nil, nil, err
+				}
+			}
+		}
+	}
+
```
Comment (eurekaka) [inline executor/analyze.go:881]: Can we combine these 2 loops for decoding into a single one?

File: executor/analyze.go [hunk 3]
```diff
@@ -748,6 +892,18 @@ func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtS
...
 
+	colLen := len(e.colsInfo)
+
+	indexPushedDownResult := <-idxNDVPushDownCh
+	if indexPushedDownResult.err != nil {
+		return 0, nil, nil, nil, nil, indexPushedDownResult.err
+	}
+	for _, offset := range indexesWithVirtualColOffsets {
+		ret := indexPushedDownResult.results[e.indexes[offset].ID]
+		rootRowCollector.NullCount[colLen+offset] = ret.Count
```
Comment (eurekaka) [inline executor/analyze.go:903]: why using the `ret.Count` as `NullCount`?

File: executor/analyze.go [hunk 4]
```diff
@@ -748,6 +892,18 @@ func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtS
 		}
 	}
 
+	colLen := len(e.colsInfo)
+
+	indexPushedDownResult := <-idxNDVPushDownCh
```
Comment (eurekaka) [inline executor/analyze.go:897]: delay this sync until we start to build stats for indexes?

File: executor/analyze.go [hunk 5]
```diff
@@ -821,14 +989,167 @@ func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtS
...
+				Tp:             tipb.AnalyzeType_TypeIndex,
+				Flags:          sc.PushDownFlags(),
+				TimeZoneOffset: offset,
+			},
+		}
+		statsVersion := new(int32)
+		*statsVersion = statistics.Version1
+		// No Top-N
+		topnSize := new(int32)
+		*topnSize = 0
```
Comment (eurekaka) [inline executor/analyze.go:1122]: why should we init these variables like this?

**Review by eurekaka — COMMENTED** [review]
this additional analyze request for virtual indexes is pretty awkward, what is the approximate accuracy loss if we simply compute the NDV from the samples?

**Comment by winoros** [comment]
Yeah, we need FM-Sketch to calculate NDV for global stats...

File: executor/analyze.go [hunk 6]
```diff
@@ -330,6 +336,35 @@ func analyzeIndexPushdown(idxExec *AnalyzeIndexExec) analyzeResult {
...
+	}
+	fms, nullHist, err := idxExec.buildSimpleStats(ranges, true)
+	if err != nil {
+		return analyzeResult{Err: err, job: idxExec.job}
+	}
+	result := analyzeResult{
+		TableID: idxExec.tableID,
+		Fms:     []*statistics.FMSketch{fms},
+		// We use histogram to get the Index's ID.
+		Hist:     []*statistics.Histogram{statistics.NewHistogram(idxExec.idxInfo.ID, 0, 0, statistics.Version1, types.NewFieldType(mysql.TypeBlob), 0, 0)},
```
[the remark this answers is not in the document]
Comment (winoros) [reply executor/analyze.go:357]: Leave me sometime to do some code refactoring later:joy:

File: executor/analyze.go (hunk 3 above)
Comment (winoros) [reply executor/analyze.go:903]: We don't collect anything except the FMSketch and null count in the request. So we use the `ret.Count` directly.

File: executor/analyze.go (hunk 2 above)
Comment (winoros) [reply executor/analyze.go:886]: Use a new function to wrap the upper one.

File: executor/analyze.go [hunk 7]
```diff
@@ -683,7 +764,50 @@ func (e *AnalyzeColumnsExec) buildResp(ranges []*ranger.Range) (distsql.SelectRe
 	return result, nil
 }
 
-func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtStats bool) (
+func (e AnalyzeColumnsExec) decodeSampleDataWithVirtualColumn(
```
Comment (qw4990) [inline executor/analyze.go:767]: Please add some comments for this function.

File: executor/analyze.go [hunk 8]
```diff
@@ -683,7 +764,50 @@ func (e *AnalyzeColumnsExec) buildResp(ranges []*ranger.Range) (distsql.SelectRe
...
+	chk := chunk.NewChunkWithCapacity(fieldTps, len(collector.Samples))
+	decoder := codec.NewDecoder(chk, e.ctx.GetSessionVars().TimeZone)
+	for _, sample := range collector.Samples {
+		for i := range sample.Columns {
+			_, err := decoder.DecodeOne(sample.Columns[i].GetBytes(), i, fieldTps[i])
+			if err != nil {
+				// Eval virtual column failed. Give up building virtual column stats.
+				logutil.BgLogger().Warn("Build stats on virtual column failed when analyzing. Stats on virtual column will be incomplete",
+					zap.Int64("table id", e.AnalyzeInfo.TableID.TableID),
+					zap.Int64("partition id", e.AnalyzeInfo.TableID.PartitionID),
```
Comment (qw4990) [inline executor/analyze.go:782]: How about printing the table name here as well?

File: executor/analyze.go [hunk 9]
```diff
@@ -524,19 +583,39 @@ func analyzeColumnsPushdown(colExec *AnalyzeColumnsExec) []analyzeResult {
...
+					isSpecial = true
+					break
+				}
+			}
+			if isSpecial {
+				specialIndexesOffsets = append(specialIndexesOffsets, i)
+				specialIndexes = append(specialIndexes, idx)
+			}
+		}
+		idxNDVPushDownCh := make(chan analyzeIndexNDVTotalResult, 1)
```
[the remark this answers is not in the document]
Comment (winoros) [reply executor/analyze.go:605]: It will a result group of all indexes.

File: executor/analyze.go [hunk 10]
```diff
@@ -785,15 +949,37 @@ func (e *AnalyzeColumnsExec) buildSamplingStats(ranges []*ranger.Range, needExtS
...
+				if col.Length != types.UnspecifiedLength {
+					row.Columns[col.Offset].Copy(&tmpDatum)
+					ranger.CutDatumByPrefixLen(&tmpDatum, col.Length, &e.colsInfo[col.Offset].FieldType)
+					b, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, b, tmpDatum)
+					if err != nil {
+						return 0, nil, nil, nil, nil, err
+					}
+					continue
+				}
 				b, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, b, row.Columns[col.Offset])
```
Comment (qw4990) [inline executor/analyze.go:986]: What if we failed to decode virtual columns and this column is virtual? Will it cause a panic?

File: executor/analyze.go [hunk 11]
```diff
@@ -512,6 +548,29 @@ func (e *AnalyzeIndexExec) buildStats(ranges []*ranger.Range, considerNull bool)
 	return hist, cms, fms, topN, nil
 }
 
+func (e *AnalyzeIndexExec) buildSimpleStats(ranges []*ranger.Range, considerNull bool) (fms *statistics.FMSketch, nullHist *statistics.Histogram, err error) {
```
Comment (time-and-fate) [inline executor/analyze.go:551]: Seems we don't need the second parament since it's always true.

**Review by qw4990 — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**